### PR TITLE
Add note about Copy for drop()

### PR DESCRIPTION
**Pull request by Manishearth**
https://users.rust-lang.org/t/confused-by-std-mem-drop/2783/3

r? @Gankro

### CODE REVIEW

**Comment by steveklabnik** [comment]
I've always been a bit wary of introducing things like this into examples, because what people get confused by is always different, and deciding what "obvious" things is hard on each example. But I'm okay with this.

File: src/libcore/mem.rs [hunk 1]
```diff
@@ -434,6 +434,10 @@ pub fn replace<T>(dest: &mut T, mut src: T) -> T {
 /// While this does call the argument's implementation of `Drop`, it will not
 /// release any borrows, as borrows are based on lexical scope.
 ///
+/// This does nothing for types which implement `Copy`, i.e. integers.
```
Comment (gkoz) [inline src/libcore/mem.rs]: "e.g. integers"

**Comment by mdinger** [comment]
@steveklabnik 
An alternative would be to add a footnote.

> [Copy types](http://doc.rust-lang.org/nightly/book/ownership.html#copy-types) may appear to misbehave but they are still following standard [move semantics](http://doc.rust-lang.org/nightly/book/ownership.html#move-semantics). It's simply the copy which follows them.

**Comment by Manishearth** [comment]
I think as far as documentation goes copy semantics have always been portrayed as being _different_ from move semantics (not "a move and then a copy")

**Comment by mdinger** [comment]
No, a copy and then a move. The original is untouched but ownership transference still applies to the new copy. Is there a reason to attempt to separate these two topics otherwise?They're on the same page and move leads directly into copy. It just cements further the idea that ownership move semantics are a pervasive idea.

**Comment by steveklabnik** [comment]
Yeah, I would say i agree with @mdinger  in the sense that we do try to make it explicit that the only difference is the ability to use it after, or at least, if we're not, we should.

**Comment by mdinger** [comment]
My preference would definitely be the footnote. Then you give them a guiding push in the correct direction and the *real* location of where this is documented. If they still don't understand, then maybe the regular docs need further improvement too.

**Comment by Manishearth** [comment]
@steveklabnik updated

File: src/libcore/mem.rs [hunk 2]
```diff
@@ -434,6 +434,11 @@ pub fn replace<T>(dest: &mut T, mut src: T) -> T {
 /// While this does call the argument's implementation of `Drop`, it will not
 /// release any borrows, as borrows are based on lexical scope.
 ///
+/// This effectively does nothing for
+/// [types which implement `Copy`](http://doc.rust-lang.org/nightly/book/ownership.html#copy-types),
+/// e.g. integers. Such values are copied and _then_ moved into the function,
+/// so the new copy persists after this function call.
```
Comment (mdinger) [inline src/libcore/mem.rs]: This disagrees slightly with the examples below where they state the copy is moved and dropped. Here it states the *the new copy persists* which is kinda the opposite. They should at least agree.

I don't know which way it's implemented or if it matters but from a explanation standpoint, letting someone copy your item so they have a new copy is more typical. It'd be more unusual if you let them copy and kept the new copy leaving them the ratty old copy you have (imagine copying homework or recipes for example).

Comment (Manishearth) [inline src/libcore/mem.rs]: fixed

File: src/libcore/mem.rs [hunk 3]
```diff
@@ -434,6 +434,11 @@ pub fn replace<T>(dest: &mut T, mut src: T) -> T {
 /// While this does call the argument's implementation of `Drop`, it will not
 /// release any borrows, as borrows are based on lexical scope.
 ///
+/// This effectively does nothing for
+/// [types which implement `Copy`](http://doc.rust-lang.org/nightly/book/ownership.html#copy-types),
```
Comment (steveklabnik) [inline src/libcore/mem.rs]: sorry, one more nit: this needs to be a relative, not absolute, link

Comment (Manishearth) [inline src/libcore/mem.rs]: done

**Pull request merged**